revocation/revocation.go

```diff
@@ -0,0 +1,122 @@
+// Package Revocation provides methods for checking the revocation status of a certificate chain
+package revocation
+
+import (
+	"crypto/x509"
+	"net/http"
+	"time"
+
+	"github.com/notaryproject/notation-core-go/revocation/ocsp"
+)
+
+// Revocation is an interface that specifies methods used for revocation checking
+type Revocation interface {
+	Validate(certChain []*x509.Certificate, signingTime time.Time) error
+	OCSPStatus(certChain []*x509.Certificate, signingTime time.Time) error
+	SetMergeErrorsFunction(mergeErrorsFunc func(errors []error) error)
+}
+
+// revocation is an internal struct used for revocation checking
+type revocation struct {
+	httpClient      *http.Client
+	mergeErrorsFunc func(errors []error) error
+}
+
+// New constructs a revocation object and substitutes default values for any that are passed as nil
+func New(httpClient *http.Client) Revocation {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+	return &revocation{
+		httpClient:      httpClient,
+		mergeErrorsFunc: DefaultMergeErrors,
+	}
+}
+
+// Validate checks OCSP, returns nil if all certs in the chain are not revoked.
+// If there is an error, it will return one of the errors defined in the revocation/ocsp package in errors.go.
+// (e.g. if a certificate in the chain is revoked by OCSP and there are no other errors, it will return ocsp.RevokedError)
+//
+// TODO: add CRL support
+func (r *revocation) Validate(certChain []*x509.Certificate, signingTime time.Time) error {
+	return r.OCSPStatus(certChain, signingTime)
+	// TODO: add CRL support
+	// https://github.com/notaryproject/notation-core-go/issues/125
+}
+
+// OCSPStatus checks OCSP, returns nil if all certs in the chain are not revoked.
+// If there is an error, it will return one of the errors defined in the revocation/ocsp package in errors.go.
+// (e.g. if a certificate in the chain is revoked by OCSP and there are no other errors, it will return ocsp.RevokedError)
+func (r *revocation) OCSPStatus(certChain []*x509.Certificate, signingTime time.Time) error {
+	return ocsp.OCSPStatus(ocsp.Options{
+		CertChain:       certChain,
+		SigningTime:     signingTime,
+		HTTPClient:      r.httpClient,
+		MergeErrorsFunc: r.mergeErrorsFunc,
+	})
+}
+
+// SetMergeErrorsFunction allows you to specify an alternative function to merge errors if the default does not fit your use case. You can also pass nil to reset it to the DefaultMergeErrors function
+func (r *revocation) SetMergeErrorsFunction(mergeErrorsFunc func(errors []error) error) {
+	if mergeErrorsFunc == nil {
+		r.mergeErrorsFunc = DefaultMergeErrors
+	} else {
+		r.mergeErrorsFunc = mergeErrorsFunc
+	}
+}
```

Comment:
Instead of setter should we take this as part of `New` ?

Reply:
My thought was this would be a lot to require of a user when they just want to get a revocation object. If they don't care about how it is merged and just want to use our default, then no work is needed. However, they can use the SetMergeErrorsFunction to have more options if they want to.

I picture this function as an extension of normal functionality to provide a better user experience, hence it is an extra function rather than a necessary incorporation to the constructor.

Reply:
One more reason why core-go shouldnot merge errors.